### PR TITLE
Add integration tests for collection estimation page

### LINKED ISSUE #893: Add integration tests for collection estimation page
The puzzle `884-d2a3ddba` from #884 has to be resolved:

https://github.com/php-coder/mystamps/blob/9287473f424222a3ead476ee7be30960c2c6dfcf/src/main/java/ru/mystamps/web/controller/CollectionController.java#L107-L107

The puzzle was created by Slava Semushin on 20-Jun-18.

---

I see that we can have at least the following 3 tests:

0) Before Test Suite:
    - login as `paid` user (with a password `test`)

1) Message should be shown when a collection is empty
    - open `/collection/paid/estimation`
    - the message "In this collection is no stamps" should be shown

2) Series with its price should be taken into account
    - open `/series/1` and it to the collection with a some price (let's use 100) and random currency
    - open `/collection/paid/estimation` page
    - the series should be listed and price should be 100 $currency (perhaps, we can use `Table Row Should Contain` for that? or `Get Table Cell`)
    - Total should be 100 $currency (perhaps, we can use `Table Footer Should Contain` for that?)

3) Series without price should be shown but not taken into account
    - open `/series/2`
    - add this series without specifying a price
    - the series should be listed without price
    - Total should be the same as before (100 $currency)

Notes:
- test cases should be placed into `src/test/robotframework/collection/estimation/logic.robot` file

### PR DESCRIPTION
**Pull request by mukeshk**
Fix #893

### CODE REVIEW

**Comment by mukeshk** [comment]
I will check. the existing test case failed.

**Comment by php-coder** [comment]
Don't forget to also remove related TODO comment (see the original issue).

File: src/test/robotframework/selenium.utils.robot [hunk 1]
```diff
@@ -1,5 +1,6 @@
 *** Settings ***
 Documentation  Keywords (and workarounds) that are missing in the SeleniumLibrary for Robot Framework
+Library    Collections
```
Comment (php-coder) [inline src/test/robotframework/selenium.utils.robot]: Please, align it properly.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 2]
```diff
@@ -0,0 +1,36 @@
+*** Settings ***
+Documentation   Verify collection estimation scenarios
+Library         SeleniumLibrary
+Resource        ../../auth.steps.robot
+Resource        ../../selenium.utils.robot
+Suite Setup     Before Test Suite
+Suite Teardown  Close Browser
+Force Tags      collection  estimation  logic  unstable
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Please, put "unstable" tag only to the cases that are unstable. For example, "Message should be shown when a collection is empty" is stable and don't need to have this tag.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+*** Test Cases ***
+Message should be shown when a collection is empty
+	Go To                   ${SITE_URL}/collection/paid/estimation
+	Element Text Should Be  id=empty-series-of-collection  In this collection is no stamps
+
+Series with its price should be taken into account
+	Go To                           ${SITE_URL}/series/1
+	Element Text Should Be          id=category_name  Prehistoric animals
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Why we need a check for category name? It seems like unrelated to what we test.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	Input Text                      id=paid-price  100
+	${currency}=                    Select Random Option From List  id=paid-currency  True
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation 
+	${amountInText}=                Catenate  100.00  ${currency}
+	Table Header Should Contain     paid-user-collection  ${amountInText}
+
+Series without price should be shown but not taken into account
+	Go To                           ${SITE_URL}/series/2
+	Element Text Should Be          id=category_name  Prehistoric animals
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Why we need a check for category name? It seems like unrelated to what we test.

File: src/test/robotframework/selenium.utils.robot [hunk 5]
```diff
@@ -69,8 +70,10 @@ Wait Until Element Value Is
...
-	[Arguments]                ${locator}
+	[Arguments]                ${locator}  ${ignoreEmptyOption}=False
 	${options}=                Get List Items  ${locator}
+	Run Keyword If             ${ignoreEmptyOption}  Remove Values From List  ${options}  ${EMPTY}
 	${size}=                   Get Length  ${options}
-	${randomIndex}=            Evaluate  random.randint(0, ${size}-1)  modules=random
-	Select From List By Index  ${locator}  ${randomIndex}
+	${randomItemIndex}=        Evaluate  random.randint(0, ${size}-1)  modules=random
+	Select From List By Value  ${locator}  @{options}[${randomItemIndex}]
+	[Return]                   @{options}[${randomItemIndex}]
```
Comment (php-coder) [inline src/test/robotframework/selenium.utils.robot]: It seems like the changes in the keyword break existing tests somehow. Also, I don't like the fact that we have to return something from the keyword.

Now I think that we should implement it a little different: in `Before Test Suite` randomly pick a currency and use it everywhere in the tests (see [`Create List`](https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Create%20List) and [`Set Suite Variable`](https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Set%20Suite%20Variable))

File: src/test/robotframework/collection/estimation/logic.robot [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	Table Header Should Contain     paid-user-collection  100.00
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: We should also check for a currency here (see the suggested approach that allow us to use a global variable).

Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: Sure will work on this approach.
In that case, should we remove the enhancement to the "selection random option from list" keyword?

File: src/test/robotframework/selenium.utils.robot (hunk 5 above)
Comment (mukeshk) [reply src/test/robotframework/selenium.utils.robot]: Get List Items was returning labels and not values.
The latest version of the library is returning the values with an optional argument "Values=True" to the keyword. So I have changed it to work on labels.

File: src/test/robotframework/collection/estimation/logic.robot (hunk 6 above)
Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: >  In that case, should we remove the enhancement to the "selection random option from list" keyword?

Yes, we should.

File: src/test/robotframework/selenium.utils.robot (hunk 5 above)
Comment (php-coder) [reply src/test/robotframework/selenium.utils.robot]: Thank you for the explanation! 👍

File: src/test/robotframework/collection/estimation/logic.robot [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	Element Text Should Be  id=empty-series-of-collection  In this collection is no stamps
+
+Series with its price should be taken into account
+	[Tags]                          unstable
+	Go To                           ${SITE_URL}/series/1
+	Input Text                      id=paid-price  100
+	Select From List By Value       id=paid-currency  ${currencyId}
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Here and below -- when defining a variable, equal sign should be close to the variable (no need an extra space).

File: src/test/robotframework/selenium.utils.robot [hunk 8]
```diff
@@ -72,5 +72,5 @@ Select Random Option From List
 	[Arguments]                ${locator}
 	${options}=                Get List Items  ${locator}
 	${size}=                   Get Length  ${options}
-	${randomIndex}=            Evaluate  random.randint(0, ${size}-1)  modules=random
-	Select From List By Index  ${locator}  ${randomIndex}
+	${randomItemIndex}=        Evaluate  random.randint(0, ${size}-1)  modules=random
+	Select From List By Index  ${locator}  ${randomItemIndex}
```
Comment (php-coder) [inline src/test/robotframework/selenium.utils.robot]: As far I understand, we don't need to modify this file anymore. Could you remove this variable renaming, then?

File: src/test/robotframework/collection/estimation/logic.robot [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	Go To                           ${SITE_URL}/series/2
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  100.00
+
+*** Keywords ***
+Before Test Suite
+	@{list} =                           Create List	USD  EUR  RUB  CZK
+	${randomCurrencyId} =               Evaluate  random.choice((${list}))  modules=random
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Why we need double braces here (`((${list}))`)?

File: src/test/robotframework/collection/estimation/logic.robot [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  100.00
+
+*** Keywords ***
+Before Test Suite
+	@{list} =                           Create List	USD  EUR  RUB  CZK
+	${randomCurrencyId} =               Evaluate  random.choice((${list}))  modules=random  
+	Set Suite Variable                  ${currencyId}  ${randomCurrencyId}
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: currencyId -> expectedCurrency

File: src/test/robotframework/collection/estimation/logic.robot (hunk 9 above)
Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: randomCurrencyId -> randomCurrency

File: src/test/robotframework/collection/estimation/logic.robot [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	[Tags]                          unstable
+	Go To                           ${SITE_URL}/series/2
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  100.00
+
+*** Keywords ***
+Before Test Suite
+	@{list} =                           Create List	USD  EUR  RUB  CZK
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: list -> currencies

File: src/main/webapp/WEB-INF/views/collection/estimation.html [hunk 12]
```diff
@@ -67,12 +67,12 @@ <h3 class="text-center" th:text="#{t_collection_of(${ownerName})}">
 					</h3>
 					
 					<!--/*/
-					<p class="text-center" th:if="${#lists.isEmpty(seriesOfCollection)}" th:text="#{t_empty_collection}">
+					<p id="empty-series-of-collection" class="text-center" th:if="${#lists.isEmpty(seriesOfCollection)}" th:text="#{t_empty_collection}">
```
Comment (php-coder) [inline src/main/webapp/WEB-INF/views/collection/estimation.html]: empty-series-of-collection -> empty-collection-msg

File: src/main/webapp/WEB-INF/views/collection/estimation.html [hunk 13]
```diff
@@ -67,12 +67,12 @@ <h3 class="text-center" th:text="#{t_collection_of(${ownerName})}">
...
 						In this collection is no stamps
 					</p>
 					/*/-->
 					
-					<table class="table table-hover" th:if="${not #lists.isEmpty(seriesOfCollection)}">
+					<table id="paid-user-collection" class="table table-hover" th:if="${not #lists.isEmpty(seriesOfCollection)}">
```
Comment (php-coder) [inline src/main/webapp/WEB-INF/views/collection/estimation.html]: paid-user-collection -> collection-estimation

File: src/test/robotframework/collection/estimation/logic.robot [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+Series with its price should be taken into account
+	[Tags]                          unstable
+	Go To                           ${SITE_URL}/series/1
+	Input Text                      id=paid-price  100
+	Select From List By Value       id=paid-currency  ${currencyId}
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  ${amountInText}
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Why we need a variable? Would it work without it? Like this:
```
Table Header Should Contain     paid-user-collection  100.00 ${currencyId}
```

File: src/test/robotframework/collection/estimation/logic.robot [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+Series without price should be shown but not taken into account
+	[Tags]                          unstable
+	Go To                           ${SITE_URL}/series/2
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  100.00
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: I see that you forgot to use amountInText variable here :) Actually, I'd suggest to inline its value as-is if it's possible.

Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: We need to add to each `Table Header Should Contain` a comment like this:
```
# TODO: use Table Footer Should Contain instead, when it will be fixed.
# See https://github.com/Hi-Fi/robotframework-seleniumlibrary-java/issues/88
```

File: src/test/robotframework/collection/estimation/logic.robot [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+Suite Setup     Before Test Suite
+Suite Teardown  Close Browser
+Force Tags      collection  estimation  logic
+
+*** Test Cases ***
+Message should be shown when a collection is empty
+	Go To                   ${SITE_URL}/collection/paid/estimation
+	Element Text Should Be  id=empty-series-of-collection  In this collection is no stamps
+
+Series with its price should be taken into account
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Please, reduce an indentation between keywords and their arguments in this test case.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+	[Tags]                          unstable
+	Go To                           ${SITE_URL}/series/1
+	Input Text                      id=paid-price  100
+	Select From List By Value       id=paid-currency  ${currencyId}
+	Submit Form                     id=add-series-form
+	Go To                           ${SITE_URL}/collection/paid/estimation
+	${amountInText} =               Catenate  100.00  ${currencyId}
+	Table Header Should Contain     paid-user-collection  ${amountInText}
+
+Series without price should be shown but not taken into account
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Please, reduce an indentation between keywords and their arguments in this test case.

File: src/test/robotframework/collection/estimation/logic.robot (hunk 15 above)
Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: image

Not working as expected

File: src/test/robotframework/collection/estimation/logic.robot (hunk 14 above)
Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: This worked

File: src/test/robotframework/collection/estimation/logic.robot (hunk 15 above)
Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: Yeah, because you don't need to use Catenate in this case:
```
Table Header Should Contain     paid-user-collection  100.00 ${currency}
```

Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: Yeah fixed it.

**Comment by mukeshk** [comment]
I am done with the comments.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+	Go To                   ${SITE_URL}/collection/paid/estimation
+	Element Text Should Be  id=empty-collection-msg  In this collection is no stamps
+
+Series with its price should be taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/1
+	Input Text                   id=paid-price  100
+	Select From List By Value    id=paid-currency  ${expectedCurrency}
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: We miss the following step: "the series should be listed and price should be 100 $currency (perhaps, we can use Table Row Should Contain for that? or Get Table Cell)".

File: src/test/robotframework/collection/estimation/logic.robot [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+	# TODO: use "Table Footer Should Contain" instead, when it will be fixed.
+    # See https://github.com/Hi-Fi/robotframework-seleniumlibrary-java/issues/88
+	Table Header Should Contain  collection-estimation  100.00 ${expectedCurrency}
+
+Series without price should be shown but not taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/2
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: We miss the following step: "the series should be listed without price"

File: src/test/robotframework/collection/estimation/logic.robot [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+Series with its price should be taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/1
+	Input Text                   id=paid-price  100
+	Select From List By Value    id=paid-currency  ${expectedCurrency}
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
+	# TODO: use "Table Footer Should Contain" instead, when it will be fixed.
+    # See https://github.com/Hi-Fi/robotframework-seleniumlibrary-java/issues/88
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: If you open this PR as a patch -- https://patch-diff.githubusercontent.com/raw/php-coder/mystamps/pull/1115.patch you will notice that there is something wrong with indentation of this line.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+	Table Header Should Contain  collection-estimation  100.00 ${expectedCurrency}
+
+Series without price should be shown but not taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/2
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
+	# TODO: use "Table Footer Should Contain" instead, when it will be fixed.
+    # See https://github.com/Hi-Fi/robotframework-seleniumlibrary-java/issues/88
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: If you open this PR as a patch -- https://patch-diff.githubusercontent.com/raw/php-coder/mystamps/pull/1115.patch you will notice that there is something wrong with indentation of this line.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+
+Series with its price should be taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/1
+	Input Text                   id=paid-price  100
+	Select From List By Value    id=paid-currency  ${expectedCurrency}
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
+	Table Cell Should Contain    collection-estimation  2  2  100.00  ${expectedCurrency}
+	Table Cell Should Contain    collection-estimation  3  2  100.00  ${expectedCurrency}
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Why we can't use `Table Header Should Contain` here? I mean, for "Total" it makes sense to use `Table Header Should Contain` while for the rows we can use `Table Cell Should Contain` or something similar.

File: src/test/robotframework/collection/estimation/logic.robot [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+	Element Text Should Be  id=empty-collection-msg  In this collection is no stamps
+
+Series with its price should be taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/1
+	Input Text                   id=paid-price  100
+	Select From List By Value    id=paid-currency  ${expectedCurrency}
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
+	Table Cell Should Contain    collection-estimation  2  2  100.00  ${expectedCurrency}
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Here and below -- first argument is locator and we should use `id=collection-estimation` to make it more obvious for a reader.

Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: It's hard to read what 2 and 2 means here. AFAIR we can use named arguments, like this:
```
Table Cell Should Contain    id=collection-estimation  row=2  col=2  text=100.00 ${expectedCurrency}
```
Could you check whether it works, please?

Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: Err.. forget my comments above. We need a better approach instead of using `Table Cell Should Contain`.

Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: > We need a better approach instead of using Table Cell Should Contain

While it's in a scope of this task, it's out of your budget, so I'm going to think about it later :)

File: src/test/robotframework/collection/estimation/logic.robot (hunk 22 above)
Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: Ok. I thought to remove the dependency on the Table Header. As we wanted to use Table Footer.
I can put it back. Yes, it more communicative and expressive in terms of Table Header.

File: src/test/robotframework/collection/estimation/logic.robot (hunk 23 above)
Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: If the table rows have id=series_x  and column has series_x_price, where x = <series_id>.
We can have an assertion with xpath - identifying tr[@id=series_1]/td[@id=series_1_price] ... kind of. Just my thought, not sure if that's your concern.

**Comment by mukeshk** [comment]
image


image

File: src/test/robotframework/collection/estimation/logic.robot (hunk 23 above)
Comment (mukeshk) [reply src/test/robotframework/collection/estimation/logic.robot]: > To summarise what needs to be done:
> 
> 1. fix "Table Cell Should Contain": add `id=` to locators, explicitly specify names of arguments (col/row/text)
> 2. reduce an indentation between keywords and their parameters in the test cases with "Table Cell Should Contain"
> 
> That's it.
> 
> Thank you for your patience, Mukesh!



> To summarise what needs to be done:
> 
> 1. fix "Table Cell Should Contain": add `id=` to locators, explicitly specify names of arguments (col/row/text)
> 2. reduce an indentation between keywords and their parameters in the test cases with "Table Cell Should Contain"
> 
> That's it.
> 
> Thank you for your patience, Mukesh!

Thanks for your patience as well. :). Let's get it right as per your expecations !

**Comment by php-coder** [comment]
In the screenshot I see that you used comma (instead of 2 spaces) to separate arguments. But you've found it already :)

File: src/test/robotframework/collection/estimation/logic.robot (hunk 23 above)
Comment (php-coder) [reply src/test/robotframework/collection/estimation/logic.robot]: > Here and below -- first argument is locator and we should use id=collection-estimation to make it more obvious for a reader.

Next time, when I'll say something wrong, you may quote documentation to me:

> Table related keywords, such as Table Should Contain, work differently. By default, when a table locator value is provided, it will search for a table with the specified id attribute.

:-)

**Comment by php-coder** [comment]
Whoops, couldn't merge because one item is still here:

> Don't forget to also remove related TODO comment (see the original issue).

Let me know when you fix it, so I'll merge.

**Comment by php-coder** [comment]
Excellent! Thank you!

File: src/test/robotframework/collection/estimation/logic.robot [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+Series with its price should be taken into account
+	[Tags]                       unstable
+	Go To                        ${SITE_URL}/series/1
+	Input Text                   id=paid-price  100
+	Select From List By Value    id=paid-currency  ${expectedCurrency}
+	Submit Form                  id=add-series-form
+	Go To                        ${SITE_URL}/collection/paid/estimation
+	Table Cell Should Contain    collection-estimation  row=2  column=2  text=100.00 ${expectedCurrency}
+	# TODO: use "Table Footer Should Contain" instead, when it will be fixed.
```
Comment (php-coder) [inline src/test/robotframework/collection/estimation/logic.robot]: Argh, I forgot that we shouldn't use TODO as 0pdd is also looking on it:

> ERROR: src/test/robotframework/collection/estimation/logic.robot; puzzle at line #23; TODO found, but puzzle can't be parsed, most probably because TODO is not followed by a puzzle marker, as this page explains: https://github.com/yegor256/pdd#how-to-format

I'll fix it later.
